### PR TITLE
[ADD][14.0] barcodes_generator_product_multi_barcode

### PR DESCRIPTION
**Pull request by AshishHirapara**
initial commit

This module expands Odoo functionality, to preserve already generated barcodes in the system.
When a barcode is generated through various other modules like (https://github.com/OCA/stock-logistics-barcode/tree/14.0/barcodes_generator_product and https://github.com/OCA/stock-logistics-barcode/tree/14.0/product_multi_barcode) and another record is already present, this module just adds it to the multi barcode table without erasing previous data.

### CODE REVIEW

File: barcodes_generator_product_multi_barcode/readme/CONFIGURE.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+No configuration needed
```
Comment (legalsylvain) [inline barcodes_generator_product_multi_barcode/readme/CONFIGURE.rst]: you can remove this file.

File: barcodes_generator_product_multi_barcode/readme/USAGE.rst [hunk 2]
```diff
@@ -0,0 +1 @@
+Just install the module
```
Comment (legalsylvain) [inline barcodes_generator_product_multi_barcode/readme/USAGE.rst]: same here.

File: barcodes_generator_product_multi_barcode/models/barcode_generate_mixin.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+    def generate_barcode(self):
+        for item in self:
+            padding = item.barcode_rule_id.padding
+            str_base = str(item.barcode_base).rjust(padding, "0")
+            custom_code = self._get_custom_barcode(item)
+            if custom_code:
+                custom_code = custom_code.replace("." * padding, str_base)
+                barcode_class = barcode.get_barcode_class(item.barcode_rule_id.encoding)
+                self.env["product.barcode"].create(
```
Comment (legalsylvain) [inline barcodes_generator_product_multi_barcode/models/barcode_generate_mixin.py:17]: Where this model is defined ?

**Review by legalsylvain — COMMENTED** [review]
Thanks for your contribution. Great feature.

Remarks inline.

Other topic : could you add tests ? 

Thanks !

File: barcodes_generator_product_multi_barcode/models/barcode_generate_mixin.py (hunk 3 above)
Comment (AshishHirapara) [reply barcodes_generator_product_multi_barcode/models/barcode_generate_mixin.py:17]: here-
https://github.com/OCA/stock-logistics-barcode/blob/9ad4e0d539c5fecfd6d1cdfbd1400937569ce27b/barcodes_generator_abstract/models/barcode_generate_mixin.py#L16

Comment (AshishHirapara) [reply barcodes_generator_product_multi_barcode/models/barcode_generate_mixin.py:17]: and the module - https://github.com/OCA/stock-logistics-barcode/blob/9ad4e0d539c5fecfd6d1cdfbd1400937569ce27b/barcodes_generator_product/__manifest__.py#L15
 is dependant on it, so I used it.

Comment (AshishHirapara) [reply barcodes_generator_product_multi_barcode/models/barcode_generate_mixin.py:17]: Regarding adding the test, the code will be covered in any of either 3 modules' test execution
https://github.com/OCA/stock-logistics-barcode/tree/14.0/barcodes_generator_product
https://github.com/OCA/stock-logistics-barcode/tree/14.0/product_multi_barcode
https://github.com/OCA/stock-logistics-barcode/tree/14.0/barcodes_generator_abstract

So still I need to write the test?

Comment (legalsylvain) [reply barcodes_generator_product_multi_barcode/models/barcode_generate_mixin.py:17]: sorry. i was talking about product.barcode model. i dont see where it is defined.

Comment (legalsylvain) [reply barcodes_generator_product_multi_barcode/models/barcode_generate_mixin.py:17]: Found it. Hère : https://github.com/OCA/stock-logistics-barcode/blob/14.0/product_multi_barcode/models/product_barcode.py

I think a dependency is missing so.

Regarding test. Yes, as your module change the behaviour of previous modules, tests will be appreciated.

Comment (AshishHirapara) [reply barcodes_generator_product_multi_barcode/models/barcode_generate_mixin.py:17]: ok, will add the test by tomorrow.

Comment (AshishHirapara) [reply barcodes_generator_product_multi_barcode/models/barcode_generate_mixin.py:17]: @legalsylvain I have added the test for my changes, but still, the tests fail here not from my module but another named 'barcodes_generator_location's'

the error log says: 
File "/__w/stock-logistics-barcode/stock-logistics-barcode/product_multi_barcode/models/product_barcode.py", line 61, in _check_duplicates
    % (record.name, barcodes[0].product_id.name)
odoo.exceptions.UserError: The Barcode "1190000000014" already exists for the product "Large Meeting Table"

There is a duplicate barcode entry constraint failing due to test data added as "1190000000014" somewhere related to barcodes_generator_location's test (https://github.com/OCA/stock-logistics-barcode/blob/d148c3cabe351ee822e057a98315cc555c8791cd/barcodes_generator_location/tests/test_barcode_generator.py#L26)

Can you suggest what should be done in this case?

Comment (legalsylvain) [reply barcodes_generator_product_multi_barcode/models/barcode_generate_mixin.py:17]: No idea. I don't use product_multi_barcode module that is a module that is changing a lot all barcode mechanism.
could you make the following test : 
- in a fresh database
- -i product_multi_barcode
- stop
- then -i barcodes_generator_location --test-enable

I suspect an error that has nothing to do with your PR.

Comment (AshishHirapara) [reply barcodes_generator_product_multi_barcode/models/barcode_generate_mixin.py:17]: Hello @legalsylvain, I have checked the test locally with your suggested params, and the test passed OK, so I don't think we need to do anything in this PR.

**Comment by francesco-ooops** [comment]
@legalsylvain is it good for you?

**Review by francesco-ooops — APPROVED** [review]
functional ok!

**Comment by legalsylvain** [comment]
Well...  I dont know how to fix the red CI...

File: barcodes_generator_product_multi_barcode/models/barcode_generate_mixin.py (hunk 3 above)
Comment (francesco-ooops) [reply barcodes_generator_product_multi_barcode/models/barcode_generate_mixin.py:17]: @AshishHirapara I see this PR: https://github.com/OCA/stock-logistics-barcode/pull/456 from 2 days ago has tests all green, so either rebase and see if tests are ok, or it means it's a problem with this module

**Pull request closed without merging**